src/thorin/continuation.cpp

```diff
@@ -11,8 +11,67 @@ namespace thorin {
 
 //------------------------------------------------------------------------------
 
-const Def* Continuation::callee() const {
-    return empty() ? world().bottom(world().fn_type(), debug()) : op(0);
+Param::Param(const Type* type, Continuation* continuation, size_t index, Debug dbg)
+    : Def(Node_Param, type, 1, dbg)
+    , index_(index)
+{
+    set_op(0, continuation);
+}
+
+//------------------------------------------------------------------------------
+
+App::App(const Defs ops, Debug dbg) : PrimOp(Node_App, ops[0]->world().bottom_type(), ops, dbg) {
+#if THORIN_ENABLE_CHECKS
+    verify();
+#endif
+}
+
+void App::verify() const {
+    auto callee_type = callee()->type()->isa<FnType>(); // works for closures too, no need for a special case
+    assertf(callee_type, "callee type must be a FnType");
+    assertf(callee_type->num_ops() == num_args(), "app node '{}' has fn type {} with {} parameters, but is supplied {} arguments", this, callee_type, callee_type->num_ops(), num_args());
+    for (size_t i = 0; i < num_args(); i++) {
+        auto pt = callee_type->op(i);
+        auto at = arg(i)->type();
+        assertf(pt == at, "app node argument {} has type {} but the callee was expecting {}", this, at, pt);
+    }
+    if (auto cont = callee()->isa_continuation()) {
+        assert(!cont->dead_);
+    }
+}
+
+const App* App::with_different_op(size_t i, const Def* d) const {
+    Array<const Def*> nops(num_ops());
+    for (size_t j = 0; j < num_ops(); j++) nops[j] = op(j);
+    nops[i] = d;
+    return world().app(nops[0], nops.skip_front(), debug());
+}
+
+const App* App::with_different_ops(const Defs defs) const {
+    return world().app(defs[0], defs.skip_front(), debug());
+}
+
+const App* App::with(const Def* ncallee, const Defs nargs) const {
+    return world().app(ncallee, nargs, debug());
+}
+
```

Comment:
do we really need this one?

Reply:
Good catch, I originally made my new nodes as Defs, not primops, and forgot to to get rid of those